src/bun.js/bindings/VMModuleScript.cpp

```diff
@@ -0,0 +1,277 @@
+        false, false, EvalContextType::None, nullptr, nullptr, ECMAMode::sloppy());
+
+    return JSValue::encode(vm.interpreter.executeEval(eval, globalThis, scope));
+}
+
+JSC_DEFINE_HOST_FUNCTION(scriptConstructorCall, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    return constructScript(globalObject, callFrame);
+}
+
+JSC_DEFINE_HOST_FUNCTION(scriptConstructorConstruct, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    return constructScript(globalObject, callFrame, callFrame->newTarget());
+}
+
+JSC_DEFINE_CUSTOM_GETTER(scriptGetCachedDataRejected, (JSGlobalObject* globalObject, EncodedJSValue thisValue, PropertyName))
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    return throwVMError(globalObject, scope, "TODO"_s);
+}
+JSC_DEFINE_HOST_FUNCTION(scriptCreateCachedData, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+    return throwVMError(globalObject, scope, "TODO"_s);
+}
+
+JSC_DEFINE_HOST_FUNCTION(scriptRunInContext, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    auto& vm = globalObject->vm();
+
+    JSValue thisValue = callFrame->thisValue();
+    auto* script = jsDynamicCast<VMModuleScript*>(thisValue);
+    if (UNLIKELY(!script)) {
+        auto scope = DECLARE_THROW_SCOPE(vm);
+        return throwVMTypeError(globalObject, scope, "Script.prototype.runInContext can only be called on a Script object"_s);
+    }
+
+    ArgList args(callFrame);
+
+    JSValue contextArg = args.at(0);
+    if (!UNLIKELY(contextArg.isObject())) {
+        auto scope = DECLARE_THROW_SCOPE(vm);
+        return throwVMTypeError(globalObject, scope, "context parameter must be a contextified object"_s);
+    }
+    JSObject* context = asObject(contextArg);
+
+    auto* zigGlobalObject = reinterpret_cast<Zig::GlobalObject*>(globalObject);
+    JSValue scopeVal = zigGlobalObject->vmModuleContextMap()->get(context);
+    if (UNLIKELY(scopeVal.isUndefined())) {
+        auto scope = DECLARE_THROW_SCOPE(vm);
+        return throwVMTypeError(globalObject, scope, "context parameter must be a contextified object"_s);
+    }
+    JSScope* scope = jsDynamicCast<JSScope*>(scopeVal);
+    ASSERT(scope);
+
+    return runInContext(globalObject, script, context, scope, args.at(1));
+}
+JSC_DEFINE_HOST_FUNCTION(scriptRunInThisContext, (JSGlobalObject* globalObject, CallFrame* callFrame))
+{
+    auto& vm = globalObject->vm();
+    JSValue thisValue = callFrame->thisValue();
+    auto* script = jsDynamicCast<VMModuleScript*>(thisValue);
+    if (UNLIKELY(!script)) {
+        auto scope = DECLARE_THROW_SCOPE(vm);
+        return throwVMTypeError(globalObject, scope, "Script.prototype.runInThisContext can only be called on a Script object"_s);
+    }
+
+    ArgList args(callFrame);
+    return runInContext(globalObject, script, globalObject->globalThis(), globalObject->globalScope(), args.at(0));
```

Comment:
We should reuse JSC's code cache like what JSC::Eval does

https://github.com/oven-sh/WebKit/blob/6bd997e38e30f50b3114b178f4d5e2fc5da6da25/Source/JavaScriptCore/interpreter/Interpreter.cpp#L152